harness/determined/common/experimental/resource_pool.py

```diff
@@ -107,3 +107,23 @@ def replace_bindings(
         bindings.put_OverwriteRPWorkspaceBindings(
             self._session, body=req, resourcePoolName=self.name
         )
+
+
+def _list(
```

Comment:
Top level, "get all the resources you know about" methods have so far lived in [determined](https://github.com/determined-ai/determined/blob/main/harness/determined/common/experimental/determined.py) and [client](https://github.com/determined-ai/determined/blob/main/harness/determined/experimental/client.py). And though the SDK is going through a bit of flux, I think we'll keep that convention for a little while.

Short note about the signature, I think ideal would be to make this public and return either a list or an iterator of `ResourcePools` rather than `bindings.v1ResourcePool`.

Reply:
In case you'd like a head start, I think my first instinct is to add attributes to `ResourcePool` objects and then give them a `to_json`. This is somewhat uncharted territory, though, so please schedule a meeting or similar with me if you think something doesn't quite feel right.

Reply:
Just for my understanding, `bindings.v1ResourcePool` is auto generated based on `message ResourcePool{}` in [resource_pool.proto](https://github.com/determined-ai/determined/blob/main/proto/src/determined/resourcepool/v1/resourcepool.proto#L66). My concern is that if we use `ResourcePools`, we will need to manually update `ResourcePools` when `message ResourcePool{}` changes.

Maybe there is a smarter way to sync `ResourcePools` and `bindings.v1ResourcePool`?

Reply:
So here, I think, is my "official" recommendation. It's based on an expectation that there are certain attributes of a resource pool that matter more than others (like you also decided on for the table), and some that we don't care about at all. And in this, Haran is undoubtably better informed than I am.

It would be nice if a `ResourcePool` was broadly useful -- if it had the functionality and attributes that users would want. Choosing the most important attributes for inclusion into the ResourcePool class and basing a `to_json` off that would help drive that broad usefulness goal.

But... I totally agree with you that it comes with the costs (that we've talked about but that I'm going to enumerate here anyway):
1. having to update the SDK's `resource_pool.py` whenever protos are updated with a new interesting attribute
2. having to choose between either giving `ResourcePool` uninteresting and irrelevant attributes or accepting that the JSONified representation will be incomplete

I mentioned in our call that I made a `Project`, and I think this is the advice I'll follow there. And I expect that at some point ml-sys will automate a better general-purpose Proto -> Python conversion layer (Anda made [MLG-809](https://hpe-aiatscale.atlassian.net/browse/MLG-809) just this morning). But there's no way, at this point, that I want it to be required for you to do the same.

Both options:
1. `list_resource_pools` in `Determined` and `client` creates `ResourcePool` objects; `to_json` in `ResourcePool`
2. `list<-whatever>` in CLI's `resource_pool.py` lists ids; `to_json` from `v1ResourcePool`

are both good options in my eyes. Please do whatever you feel best about.

Brief aside that applies to either (1) or (2), we've so far exposed/encouraged _very_ little pagination (even though the master supports it). I recommend omitting limit from the call to master if the user doesn't provide one, and considering even not making it a part of the interface entirely (to keep the interface simple until/unless there's a voiced need to do otherwise). I even more strongly suggest that `offset` not be a part of this interface for the same reason.

Reply:
Hope this is okay. `list_resource_pools` in CLI's `resource_pool.py` lists rp with attributes suggested by Haran; to_json in `ResourcePool`.

I didn't implement it exactly the same as the options you mentioned above for the following reasons:
Option 1: Call `Determined` or `client` functions in CLI's `resource_pool.py` is awkward. [They use different ways to authenticate.](https://github.com/determined-ai/determined/pull/7539/files#r1286345545)
Option 2: I may misunderstand your point. I feel it's more useful if we return more attributes than only returning ids.

Reply:
Option 1: I wrote https://github.com/determined-ai/determined/pull/7663 to make this easier, but it's been a really busy week and I didn't get it merged. I'm out until Tuesday, but will try again when I get back.

Option 2: I don't understand my point there now, either... It's too specific to have been a typo, right? Maybe it was a typo.